File: windows/tests/ConnectorControl.Core.Tests/TestSupport/AppStateHarness.cs
using ConnectorControl.Core.State;

namespace ConnectorControl.Core.Tests.TestSupport;

/// <summary>
/// A real on-disk layout (%LOCALAPPDATA% and %APPDATA% under one temp dir),
/// the real path resolver, real ConfigService/FileWatcher, and fakes for the
/// platform interfaces only. Marshal is a queue: watcher callbacks reach state
/// only when a test pumps, mirroring the UI thread.
/// </summary>
public sealed class AppStateHarness : IDisposable
{
    public TempDir Dir { get; } = new("appstate");
    public string Local => Dir.File("Local");
    public string Roaming => Dir.File("Roaming");
    public string ClaudeConfigPath => Path.Combine(Roaming, "Claude", "claude_desktop_config.json");
    public string StoreDir => Path.Combine(Local, AppPaths.DataDirName);
    public string MasterStorePath => Path.Combine(StoreDir, "mcps.json");
    public string BackupsDir => Path.Combine(StoreDir, "backups");

    public FakeSettings Settings { get; } = new();
    public FakeClaudeProcess Claude { get; } = new();
    public FakeNotifier Notifier { get; } = new();
    public FakeDialogs Dialogs { get; } = new();
    public FakeToolProbe Tools { get; } = new();
    public DelayQueue Delays { get; } = new();
    public MarshalQueue Ui { get; } = new();
    private readonly List<CollectionsModel> models = [];
    public DateTime Now { get; set; } = new(2026, 9, 4, 12, 0, 0, DateTimeKind.Utc);
    public PathContext Context { get; }
    public AppHost Host { get; }

    public AppStateHarness(bool seedClaudeConfig = true)
    {
        Directory.CreateDirectory(Path.Combine(Roaming, "Claude"));
        Directory.CreateDirectory(Local);
        if (seedClaudeConfig)
        {
            File.WriteAllText(ClaudeConfigPath, Fixtures.RealisticClaudeConfig);
        }
        Context = new PathContext(new Dictionary<string, string>(StringComparer.Ordinal), new KnownFolders(Local, Roaming), new RealPathProbe());
        Host = new AppHost(Ui.Post, Delays.Add, () => Now);
    }

    public AppState Create() => new(Settings, Claude, Notifier, Dialogs, Context, Host, Tools);

    public IReadOnlyDictionary<string, JsonValue> ClaudeServers() => ClaudeConfigIO.ReadMcpServers(ClaudeConfigPath);

    public MasterStore StoreOnDisk() => MasterStoreIO.Read(MasterStorePath) ?? throw new InvalidOperationException("no readable store on disk");

    /// <summary>An "external" edit of Claude's config: replaces mcpServers, keeps every other key.</summary>
    public void WriteClaudeServers(params (string Name, JsonValue Config)[] servers) =>
        ClaudeConfigIO.Write(servers.ToDictionary(s => s.Name, s => s.Config, StringComparer.Ordinal), ClaudeConfigPath);

    public static JsonValue Remote(string url) => RemotePattern.Make(url, RemoteLaunchStyle.Npx);

    /// <summary>A local connector running <paramref name="command"/> with <paramref name="args"/>.</summary>
    public static McpEntry LocalConnector(string command, params string[] args) =>
        new(JsonValue.Object(
            ("command", JsonValue.String(command)),
            ("args", JsonValue.Array(args.Select(JsonValue.String)))));

    public static string[] Keys(IEnumerable<string> keys) => keys.Order(StringComparer.Ordinal).ToArray();

    /// <summary>Claude running since <paramref name="hours"/> before the harness clock, so the next
    /// apply calls for a restart.</summary>
    public void ClaudeRunningSince(double hours)
    {
        Claude.IsRunning = true;
        Claude.LaunchDate = Now.AddHours(-hours);
    }

    // MARK: collections

    /// <summary>The bytes an author's machine would have written, at <paramref name="name"/> under
    /// the temp dir.</summary>
    public string WriteDocument(CollectionDocument doc, string name)
    {
        var path = Dir.File(name);
        WriteDocumentAt(doc, path);
        return path;
    }

    /// <summary>The same, at a path the test already holds: the author's next commit to a
    /// document.</summary>
    public static void WriteDocumentAt(CollectionDocument doc, string path)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(path)!);
        File.WriteAllBytes(path, doc.Serialize());
    }

    /// <summary>Writes <paramref name="doc"/> at <paramref name="name"/> under the temp dir and
    /// subscribes <paramref name="state"/> to it, as <paramref name="collection"/> or under the
    /// document's own name; a refusal fails the test. Returns where the document is.</summary>
    public string Subscribe(AppState state, CollectionDocument doc, string name = "data-team.json", string? collection = null)
    {
        var path = WriteDocument(doc, name);
        Assert.Null(state.Subscribe(path, collection));
        return path;
    }

    /// <summary>Writes both collection files where the app reads them, then reloads so the state
    /// picks them up — the shape a subscribe or a publish would leave behind.</summary>
    public void Seed(AppState state, CollectionsFile file, CollectionsLocalCache? cache = null)
    {
        file.Save(Path.Combine(StoreDir, CollectionsFile.FileName));
        (cache ?? new CollectionsLocalCache([], [])).Save(state.Service.Paths.CollectionsCachePath);
        state.Reload();
    }

    /// <summary>Marks <paramref name="name"/>, already in the store, synced from
    /// <c>&lt;slug&gt;.json</c>: found at <paramref name="boundTo"/> on this machine or, with null,
    /// not found here. It is the only collection the sidecar describes and the only synced binding
    /// in the cache; the rest of the cache is kept.</summary>
    public void MakeSynced(AppState state, string name, string? boundTo = null)
    {
        var entry = new CollectionsFile.Entry(CollectionKind.Synced, Slug.Make(name) + ".json");
        var bindings = new Dictionary<string, CollectionsLocalCache.SyncedBinding>(StringComparer.Ordinal);
        if (boundTo is not null)
        {
            bindings[name] = new CollectionsLocalCache.SyncedBinding(boundTo, null, []);
        }
        Seed(state, new CollectionsFile([new(name, entry)]), state.CollectionsCache with { Synced = bindings });
    }

    /// <summary>Publishes <paramref name="collection"/> into a new folder <paramref name="folder"/>
    /// under the temp dir; a refusal fails the test. Returns the document it wrote.</summary>
    public string Publish(AppState state, string collection, PublishIntent? intent = null, string folder = "pub")
    {
        var path = Dir.File(folder);
        Directory.CreateDirectory(path);
        Assert.Null(state.StartPublishing(collection, path, intent ?? PublishIntent.None));
        return Path.Combine(path, CollectionDocument.FileName(Slug.Make(collection)));
    }

    /// <summary>Reads the store off disk, lets <paramref name="edit"/> change it and saves it back:
    /// a write from another machine or an older app, which the running state has not seen.</summary>
    public void EditStoreOnDisk(Action<MasterStore> edit)
    {
        var store = StoreOnDisk();
        edit(store);
        MasterStoreIO.Save(store, MasterStorePath);
    }

    /// <summary>A Collections window's model on <paramref name="state"/>, showing
    /// <paramref name="selecting"/> (the active collection when null) with
    /// <paramref name="ticking"/> ticked. The harness disposes it.</summary>
    public CollectionsModel CollectionsModel(AppState state, string? selecting = null, params string[] ticking)
    {
        var model = new CollectionsModel(state, Dialogs);
        models.Add(model);
        if (selecting is not null)
        {
            model.Selected = selecting;
        }
        foreach (var name in ticking)
        {
            model.SetChecked(name, true);
        }
        return model;
    }

    public void Dispose()
    {
        foreach (var model in models)
        {
            model.Dispose();
        }
        models.Clear();
        Dir.Dispose();
    }
}

File: Tests/ConnectorControlStateTests/TestSupport/AppStateHarness.swift
import Foundation
import XCTest
import ConnectorControlCore
import ConnectorControlTestSupport
@testable import ConnectorControlState

/// A real on-disk layout (a throwaway ~/Library/Application Support), the real
/// path rule, real ConfigService and FileWatchers, and fakes for the platform
/// seams only. Marshal is a queue: watcher callbacks and probe results reach
/// state only when a test pumps, mirroring the main actor in the app.
@MainActor
final class AppStateHarness {
    struct HarnessError: Error {}

    let dir = TempDir(prefix: "appstate")
    var appSupport: URL { dir.file("Library/Application Support") }
    var claudeConfigURL: URL { appSupport.appendingPathComponent("Claude/claude_desktop_config.json") }
    var storeDir: URL { appSupport.appendingPathComponent("Connector Control") }
    var masterStoreURL: URL { storeDir.appendingPathComponent("mcps.json") }
    var backupsDir: URL { storeDir.appendingPathComponent("backups") }
    var stagingDir: URL { storeDir.appendingPathComponent(".staging") }

    let settings = FakeSettings()
    let claude = FakeClaudeProcess()
    let notifier = FakeNotifier()
    let dialogs = FakeDialogs()
    let tools = FakeToolProbe()
    let delays = DelayQueue()
    let ui = MarshalQueue()
    /// The clock every lastApplyDate is stamped with; tests move it.
    var now = ISO8601DateFormatter().date(from: "2026-09-04T12:00:00Z")!
    private var created: [AppState] = []
    private var models: [CollectionsModel] = []

    var context: PathContext { PathContext(environment: [:], appSupport: appSupport) }

    var host: AppHost {
        AppHost(marshal: { [ui] in ui.post($0) }, delay: { [delays] in delays.add($0, $1) },
                now: { [unowned self] in MainActor.assumeIsolated { self.now } })
    }

    /// `createClaudeDirectory: false` leaves even the Claude folder absent, so the
    /// Claude-config watcher cannot arm at launch (the re-arm test).
    init(seedClaudeConfig: Bool = true, createClaudeDirectory: Bool = true) {
        let fm = FileManager.default
        try! fm.createDirectory(at: appSupport, withIntermediateDirectories: true)
        if createClaudeDirectory {
            try! fm.createDirectory(at: claudeConfigURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        }
        if seedClaudeConfig {
            try! Data(Fixtures.realisticClaudeConfig.utf8).write(to: claudeConfigURL)
        }
    }

    func create() -> AppState {
        let state = AppState(settings: settings, claude: claude, notifier: notifier, dialogs: dialogs,
                             paths: context, host: host, toolProbe: tools)
        created.append(state)
        return state
    }

    func claudeServers() throws -> [String: JSONValue] {
        try ClaudeConfigIO.readMCPServers(at: claudeConfigURL)
    }

    func storeOnDisk() throws -> MasterStore {
        guard let store = MasterStoreIO.read(from: masterStoreURL) else { throw HarnessError() }
        return store
    }

    /// An "external" edit of Claude's config: replaces mcpServers, keeps every other key.
    func writeClaudeServers(_ servers: [(String, JSONValue)]) throws {
        try ClaudeConfigIO.write(mcpServers: Dictionary(uniqueKeysWithValues: servers), to: claudeConfigURL)
    }

    static func remote(_ url: String) -> JSONValue {
        RemotePattern.make(url: url)
    }

    /// A local connector running `command` with `args`.
    static func localConnector(_ command: String, _ args: [String] = []) -> MCPEntry {
        MCPEntry(config: .object(["command": .string(command), "args": .array(args.map(JSONValue.string))]))
    }

    /// Claude running since `hours` before the harness clock, so the next apply calls for a restart.
    func claudeRunningSince(hours: Double) {
        claude.isRunning = true
        claude.launchDate = now.addingTimeInterval(-hours * 3600)
    }

    // MARK: Collections

    /// The bytes an author's machine would have written, at `name` under the temp dir.
    @discardableResult
    func writeDocument(_ doc: CollectionDocument, named name: String) throws -> URL {
        let url = dir.file(name)
        try writeDocument(doc, at: url)
        return url
    }

    /// The same, at a path the test already holds: the author's next commit to a document.
    func writeDocument(_ doc: CollectionDocument, at url: URL) throws {
        try FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try doc.serialized().write(to: url)
    }

    /// Writes `doc` at `name` under the temp dir and subscribes `state` to it, as `collection` or
    /// under the document's own name; a refusal fails the test. Returns where the document is.
    @discardableResult
    func subscribe(_ state: AppState, to doc: CollectionDocument, at name: String = "data-team.json",
                   as collection: String? = nil, file: StaticString = #filePath, line: UInt = #line) throws -> URL {
        let url = try writeDocument(doc, named: name)
        XCTAssertNil(state.subscribe(documentAt: url.path, as: collection), file: file, line: line)
        return url
    }

    /// Writes both collection files where the app reads them, then reloads so the state picks
    /// them up — the shape a subscribe or a publish would leave behind.
    func seed(_ state: AppState, file: CollectionsFile, cache: CollectionsLocalCache? = nil) throws {
        try file.save(to: storeDir.appendingPathComponent(CollectionsFile.fileName), staging: nil)
        try (cache ?? CollectionsLocalCache(synced: [:], published: [:]))
            .save(to: state.service.paths.collectionsCacheURL, staging: nil)
        state.reload()
    }

    /// Marks `name`, already in the store, synced from `<slug>.json`: found at `path` on this
    /// machine or, with nil, not found here. It is the only collection the sidecar describes and
    /// the only synced binding in the cache; the rest of the cache is kept.
    func makeSynced(_ state: AppState, _ name: String, boundTo path: String? = nil) throws {
        let entry = CollectionsFile.Entry(kind: .synced, fileName: Slug.make(name) + ".json")
        var cache = state.collectionsCache
        cache.synced = path.map { [name: CollectionsLocalCache.SyncedBinding(path: $0, lastHash: nil, excluded: [:])] } ?? [:]
        try seed(state, file: CollectionsFile(collections: [name: entry]), cache: cache)
    }

    /// Publishes `collection` into a new folder `folder` under the temp dir; a refusal fails the
    /// test. Returns the document it wrote.
    @discardableResult
    func publish(_ state: AppState, _ collection: String, intent: PublishIntent = .none, folder: String = "pub",
                 file: StaticString = #filePath, line: UInt = #line) throws -> URL {
        let url = dir.file(folder)
        try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        XCTAssertNil(state.startPublishing(collection, to: url.path, intent: intent), file: file, line: line)
        return url.appendingPathComponent(CollectionDocument.fileName(slug: Slug.make(collection)))
    }

    /// Reads the store off disk, lets `edit` change it and saves it back: a write from another
    /// machine or an older app, which the running state has not seen.
    func editStoreOnDisk(_ edit: (inout MasterStore) throws -> Void) throws {
        var store = try storeOnDisk()
        try edit(&store)
        try MasterStoreIO.save(store, to: masterStoreURL)
    }

    /// A Collections window's model on `state`, showing `selecting` (the active collection when
    /// nil) with `ticking` ticked. The harness disposes it.
    func collectionsModel(_ state: AppState, selecting: String? = nil, ticking: [String] = []) -> CollectionsModel {
        let model = CollectionsModel(state: state, dialogs: dialogs)
        models.append(model)
        if let selecting { model.selected = selecting }
        for name in ticking { model.setChecked(name, true) }
        return model
    }

    /// Disposes every model and AppState this harness created (stops their watchers) and deletes
    /// the temp dir.
    func dispose() {
        models.forEach { $0.dispose() }
        models.removeAll()
        created.forEach { $0.dispose() }
        created.removeAll()
        dir.dispose()
    }

    /// The harness plus its AppState, in one call — the shape most tests
    /// need. `let (h, state) = AppStateHarness.started(); defer { h.dispose() }`.
    /// A test that must act on the harness BEFORE the AppState exists (seed a
    /// fake's state, change a setting) still uses `AppStateHarness()` +
    /// `h.create()` directly.
    static func started(seedClaudeConfig: Bool = true, createClaudeDirectory: Bool = true)
        -> (AppStateHarness, AppState) {
        let h = AppStateHarness(seedClaudeConfig: seedClaudeConfig, createClaudeDirectory: createClaudeDirectory)
        return (h, h.create())
    }

    /// Advances the mtime of Claude's config and the master store (whichever
    /// currently exist) past whatever they currently are — called after an
    /// action (direct or through AppState) that must be observably distinct,
    /// to the watchers, from whatever they last saw at arm time. Replaces
    /// `Thread.sleep` used only to separate two writes' mtimes by real
    /// wall-clock time.
    func touchWatchedFiles() throws {
        let fm = FileManager.default
        if fm.fileExists(atPath: claudeConfigURL.path) {
            try TempDir.bumpModificationDate(of: claudeConfigURL)
        }
        if fm.fileExists(atPath: masterStoreURL.path) {
            try TempDir.bumpModificationDate(of: masterStoreURL)
        }
    }
}
